### PR TITLE
[RUST] Flat index benchmark

### PR DESCRIPTION
**Pull request by eddyxu**
* Provide a benchmark for `Flat` index
* Implement Apple Accelerate / Blas and NEON SIMD versions of L2 distance
* Currently BLAS only works on [MacOS + Accelerate](https://developer.apple.com/documentation/accelerate?language=objc)

### CODE REVIEW

**Comment by eddyxu** [comment]
Flat index run on SIFT-1M dataset, on a M1 Macbook Air 16GB

With Blas / Apple Accelerate Framework
Screenshot 2023-01-25 at 3 19 28 PM

With hand written ARM NEON SIMD

Screenshot 2023-01-25 at 3 22 35 PM

File: rust/src/utils/distance.rs [hunk 1]
```diff
@@ -60,12 +61,64 @@ unsafe fn euclidean_distance_fma(from: &[f32], to: &[f32]) -> f32 {
 ///
 #[inline]
 pub fn l2_distance_arrow(from: &Float32Array, to: &Float32Array) -> f32 {
-    let mul: Float32Array = binary(from, to, |a, b| (a - b).powf(2.0)).unwrap();
-    sum(&mul).unwrap()
+    let a = from.values();
+    let b = to.values();
+    let mut d = 0.0;
+    // Better chance to auto-vectorization.
```
Comment (eddyxu) [inline rust/src/utils/distance.rs:67]: Only got about 5% from this.

Comment (changhiskhan) [reply rust/src/utils/distance.rs:67]: that's interesting. thought those blog posts showed bigger gains

**Review by changhiskhan — APPROVED** [review]
how do we distribute this in the python package?

**Comment by eddyxu** [comment]
>how do we distribute this in the python package?

For macOS, we can build it with `cargo release --features="blas"`. The other platform, release as old as usual. 

> that's interesting. thought those blog posts showed bigger gains

I think i did not try hard enough.

**Pull request merged**